### PR TITLE
Retry transaction sending when gas prices increase

### PR DESCRIPTION
**Pull request by vkgnosis**
Using the retry crate from gpv1.

A lot of the code is similar to solution_submission.rs in gp-v1 but adapted to the rest of the gpv2 code. `is_transaction_error` and `GasPriceStream` are exactly the same. It wasn't worth to factor those into individual crates because they are small and the stream relies on a sleep trait which is project specific.

### Test Plan
CI e2e test.

### CODE REVIEW

File: e2e/tests/uniswap_trade_test.rs [hunk 1]
```diff
@@ -59,7 +59,13 @@ async fn test_with_ganache() {
     let uniswap_router = UniswapV2Router02::deployed(&web3)
         .await
         .expect("Failed to load deployed UniswapFactory");
-    let gp_settlement = GPv2Settlement::deployed(&web3)
+
+    // I took this private key from the gp-v1 e2e test. Not sure why this actually works.
+    // Does Ganache guarantee that it works?
+    let private_key = "4f3edf983ac636a65a842ce7c78d9aa706d3b113bce9c46f30d7d21715b23b1d"
```
Comment (fleupold) [inline e2e/tests/uniswap_trade_test.rs:65]: If ganache is started with the default seed phrase, this will be the 0th account in the list. I don't think we should rely on this though. The more general way to use accounts is via the `Accounts` type. On ganache accounts are "unlocked" meaning they can sign transactions. For xDAI/mainnet/Rinkeby we can create an offline account that is initialized with a private key.

I think `get_settlement_contract` should take `Account` as an argument instead of private key, so that we can pass account[0] in the e2e test.

File: solver/src/driver.rs [hunk 2]
```diff
@@ -80,15 +84,29 @@ impl Driver {
...
             self.target_confirm_time,
+            GAS_PRICE_CAP,
+            settlement,
         )
         .await
         .context("failed to submit settlement")?;
         Ok(())
     }
+
+    async fn transaction_count(&self) -> Result<U256> {
```
Comment (fleupold) [inline solver/src/driver.rs:105]: nit: why have this logic here and not in `submit`? Seems like it would fit more coherently over there.

File: solver/src/settlement_submission/retry.rs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+// Create a never ending stream of gas prices based on checking the estimator in fixed intervals
+// and enforcing the minimum increase. Errors are ignored.
+pub fn gas_price_stream(
+    target_confirm_time: Duration,
+    gas_price_cap: f64,
+    estimator: &dyn GasPriceEstimating,
+) -> impl Stream<Item = f64> + '_ {
+    let stream = stream::unfold(true, move |first_call| async move {
+        if !first_call {
+            tokio::time::delay_for(GAS_PRICE_REFRESH_INTERVAL).await;
```
Comment (fleupold) [inline solver/src/settlement_submission/retry.rs:26]: ubernit: move constant to this module?

File: solver/src/settlement_submission/retry.rs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        }
+        let estimate = estimator
+            .estimate_with_limits(MAX_GAS as f64, target_confirm_time)
```
Comment (fleupold) [inline solver/src/settlement_submission/retry.rs:29]: Can we add a todo to use the real estimate gas amount instead of MAX_GAS?

File: solver/src/settlement_submission/retry.rs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+            Err(err) => {
+                tracing::error!("gas price estimation failed: {:?}", err);
+                None
+            }
+        }
+    });
+    transaction_retry::gas_price_increase::enforce_minimum_increase_and_cap(gas_price_cap, stream)
+}
+
+fn is_transaction_error(error: &ExecutionError) -> bool {
```
Comment (fleupold) [inline solver/src/settlement_submission/retry.rs:48]: It's a shame we are duplicating this logic. Would it not make sense to also expose this method in the retry crate (not blocking this PR but generally)?

File: solver/src/settlement_submission/retry.rs [hunk 6]
```diff
@@ -0,0 +1,136 @@
+use super::{EncodedSettlement, GAS_PRICE_REFRESH_INTERVAL, MAX_GAS};
```
Comment (fleupold) [inline solver/src/settlement_submission/retry.rs:1]: It feels to me as if this module is collecting a bunch of parts that are not really related to one another. Maybe the gas_price_stream should be part of the `gas_price_estimation` module and SettleResult/SettlementSender also seem not strictly related to retry. Maybe this module could be called transaction?

File: solver/src/settlement_submission/retry.rs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+                tracing::debug!("estimated gas price {}", gas_price);
+                Some(gas_price)
+            }
+            Err(err) => {
+                tracing::error!("gas price estimation failed: {:?}", err);
+                None
+            }
+        }
+    });
+    transaction_retry::gas_price_increase::enforce_minimum_increase_and_cap(gas_price_cap, stream)
```
Comment (fleupold) [inline solver/src/settlement_submission/retry.rs:45]: I personally think it would be nice to expose a gas price stream factory that takes a target time, amount, price_cap, estimator and a "sleep" function from the retry or price estimating crate. It seems generic enough to me and the logic with the unfolded stream is a bit nifty (so nice not to maintain it twice).

File: solver/src/settlement_submission/retry.rs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+        tracing::info!("submitting solution transaction at gas price {}", gas_price);
+        let mut method = self
+            .contract
+            .settle(
+                self.settlement.tokens.clone(),
+                self.settlement.clearing_prices.clone(),
+                self.settlement.encoded_trades.clone(),
+                self.settlement.encoded_interactions.clone(),
+                Vec::new(),
+            )
```
Comment (fleupold) [inline solver/src/settlement_submission/retry.rs:85]: nit: This code is duplicated between simulate and execute. Maybe the settlement helper struct can expose a function that returns a method builder given?

**Review by fleupold — APPROVED** [review]
The only blocking thing is that I wouldn't rely on ganache being started with the correct seed phrase. Otherwise mostly nits, I think it would have been nice to reuse more of the logic, but ok.

File: solver/Cargo.toml [hunk 9]
```diff
@@ -17,6 +17,7 @@ async-trait = "0.1"
 contracts = { path = "../contracts" }
 ethcontract = { version = "0.11", default-features = false }
 futures = "0.3"
+transaction-retry = { git = "https://github.com/gnosis/gp-transaction-retry.git", rev = "b12c37e0916854dbf046d14e17cc4a7765666610" }
```
Comment (bh2smith) [inline solver/Cargo.toml:20]: Have we not settled on a version of this yet?

File: solver/src/driver.rs [hunk 10]
```diff
@@ -7,10 +7,14 @@ use crate::{
...
 use contracts::GPv2Settlement;
 use gas_estimation::GasPriceEstimating;
+use primitive_types::U256;
 use std::time::Duration;
 use tracing::info;
 
 const SETTLE_INTERVAL: Duration = Duration::from_secs(30);
+// There is no economic viability calculation yet so we're using an arbitrary very high cap to
+// protect against a gas estimator giving bogus results that would drain all our funds.
+const GAS_PRICE_CAP: f64 = 500e9;
```
Comment (bh2smith) [inline solver/src/driver.rs:16]: Seems reasonable... 500 Gwei happens sometimes.

File: solver/src/driver.rs (hunk 2 above)
Comment (bh2smith) [reply solver/src/driver.rs]: I am actually surprised how many lines are needed just to get the nonce. Also, not sure what the settlement contract has to do with an EOA's nonce.

**Review by bh2smith — COMMENTED** [review]
Looks nice so far. Seems there are a few things to cleanup at this time.

**Review by bh2smith — APPROVED** [review]

File: solver/src/settlement_submission/retry.rs [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use jsonrpc_core::ErrorCode;
+
+    #[test]
+    fn test_submission_result_was_mined() {
```
Comment (bh2smith) [inline solver/src/settlement_submission/retry.rs:118]: Does this test somehow include a retry? If so, I am unable to tell how.

File: solver/src/settlement_submission/retry.rs (hunk 6 above)
Comment (vkgnosis) [reply solver/src/settlement_submission/retry.rs]: Moved the stream to it's own module. Don't like it in gas_price_estimation because that is just responsible for creating the estimator whereas the stream is only used for retry.

Comment (vkgnosis) [reply solver/src/settlement_submission/retry.rs]: > SettleResult/SettlementSender also seem not strictly related to retry. Maybe this module could be called transaction

Imo they are because these types are only used for retry.

File: solver/src/settlement_submission/retry.rs (hunk 7 above)
Comment (vkgnosis) [reply solver/src/settlement_submission/retry.rs]: The problem is that there are too many implementation details like sleeping, logging and how to call the estimator and there is no good generic way to take them. To me duplicating makes more sense.

File: solver/src/settlement_submission/retry.rs (hunk 5 above)
Comment (vkgnosis) [reply solver/src/settlement_submission/retry.rs:15]: I don't think it makes sense in the retry crate because it is ethcontract specific. Maybe behind an ethcontract feature but then it is annoying when the version of ethcontract doesn't match. Maybe it would make sense in ethcontract or web3 directly but I don't think so because it is so node specific.

File: solver/src/settlement_submission/retry.rs (hunk 3 above)
Comment (vkgnosis) [reply solver/src/settlement_submission/retry.rs]: Used in several places now so should stay in parent.

File: solver/src/settlement_submission/retry.rs (hunk 11 above)
Comment (vkgnosis) [reply solver/src/settlement_submission/retry.rs:91]: it does not and is not supposed to. just tests that detection of was_mined works.

File: solver/Cargo.toml (hunk 9 above)
Comment (vkgnosis) [reply solver/Cargo.toml:20]: I haven't pushed a tag for it yet so no. When we make the next change in that repo we can set a version.

File: solver/src/settlement_submission/retry.rs (hunk 7 above)
Comment (fleupold) [reply solver/src/settlement_submission/retry.rs]: This discussion goes outside the scope of this PR, but what's the concern of making the retry library opinionated on e.g. logging/sleeping (not that we could pass it as abstract traits from above) similar to how other libraries (e.g. web3 are)? 

It find there is still quite a lot of manual code involved to integrate transaction retries and it feels to me that the use case of wanting to submit some MethodBuilder with a specific target time and a gas price estimator is quite common.

Comment (vkgnosis) [reply solver/src/settlement_submission/retry.rs]: I suspect that it wouldn't be useful because people want to customize too many parts of it. Like what if you want to do some extra work every time a resubmission happens other than what it provides by default. I did make an issue now on the retry repo to try implementing this with an optional `ethcontract` feature so I am going to try it out.
I have another idea for the gas price stream too.
